caffe2/contrib/nnpack/nnpack_ops.cc

```diff
@@ -335,9 +335,38 @@ class NNPACKLeakyReluOp final : public LeakyReluOp<float, CPUContext> {
  private:
 };
 
+class NNPACKSoftmaxOp final : public Operator<CPUContext> {
+ public:
+  NNPACKSoftmaxOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<CPUContext>(operator_def, ws) {
+#ifdef CAFFE2_USE_FBCODE
+    // Facebook's nnpack build assumes existence of avx2, so we explicitly
+    // check if the machine has avx2 support.
+    OPERATOR_NEEDS_FEATURE(GetCpuId().avx2(), "NNPack requires AVX2");
+#endif
+  }
+
+  bool RunOnDevice() override {
+    auto& X = Input(0);
+    const auto N = X.dim32(0), C = X.dim32(1), H = X.dim32(2), W = X.dim32(3);
+    auto* Y = Output(0);
+    auto channels = C * H * W;
```

Comment:
This is actually not right - see https://github.com/caffe2/caffe2/blob/master/caffe2/operators/softmax_op.cc#L9-L13 for the right index calculation.